### ISSUE TITLE
Consul ACL Policy - k8s to Consul

### ISSUE THREAD

**Issue by arctiqtim** [hashicorp/consul-k8s#26]
On a Consul setup with default deny policy and ACLs enabled I can't seem to get the correct policy to allow the service sync to work from k8s to consul.

Using a token with the following policy:
```
    node_prefix "" {
      policy = "write"
    }
    service_prefix "" {
      policy = "read"
    }
```
I see the following errors:
```
    2018/10/29 02:50:24 [ERR] consul: "Catalog.Register" RPC failed to server x.x.x.x:8300: rpc error making call: rpc error making call: Permission denied
    2018/10/29 02:50:24 [ERR] http: Request PUT /v1/catalog/register, error: rpc error making call: rpc error making call: Permission denied from=x.x.x.x:38006
```

Relaxing the policy with:
```
    node_prefix "" {
      policy = "write"
    }
    service_prefix "" {
      policy = "read"
    }
    agent_prefix "" {
      policy = "write"
    }
    session_prefix "" {
      policy = "write"
    }
```
I am seeing the following errors:
```
2018-10-29T03:07:02.848Z [WARN ] to-consul/sink: error registering service: node-name=k8s-sync service-name=test err="Unexpected response code: 403 (rpc error making call: rpc error making call: Permission denied)"
```

If I change the consul-k8s deployment to use the master token instead, the sync'ing services work perfectly.  Is there any guidance on what ACL policies are required in order for this sync to work?

This is currently running in GKE with the latest Consul release, so the newer ACL scheme.
```/ # consul members
Node                 Address         Status  Type    Build     Protocol  DC   Segment
consul-0             x.x.x.x:8301  alive   server  1.4.0rc1  2         cs1  <all>
consul-1             x.x.x.x:8301  alive   server  1.4.0rc1  2         cs1  <all>
consul-2             x.x.x.x:8301  alive   server  1.4.0rc1  2         cs1  <all>
consul-agents-fzl8j  x.x.x.x:8301  alive   client  1.4.0rc1  2         cs1  <default>
consul-agents-g6l9d  x.x.x.x:8301  alive   client  1.4.0rc1  2         cs1  <default>
consul-agents-xdfl7  x.x.x.x:8301  alive   client  1.4.0rc1  2         cs1  <default>
```

**Comment by dedene** [comment]
In case it helps, with ACLs enabled I also needed to give the anonymous token a write access policy for both bodes and services. After that the Permission denied errors were gone.

**Comment by banks** [comment]
@arctiqtim the policy you need for two way sync should be:

```
# read all nodes allows the sync process to _discover_ services to sync them 
# _to_ k8s since it needs read on the node that they are one to be able to "see" them
node_prefix "" {
  policy = "read"
}

# write on all services allows the sync process to create service registrations for
# anything it finds in k8s. If you are using -k8s-service-prefix then you can lock this 
# down further by only granting write for that specific prefix which is safer since 
# service:write grants access to the _identity_ of that service and hence anything it has
# access to when using Connect.
service_prefix "" {
  policy = "write"
}
```

I've not actually tested that so there might be some other subtelty but those are the main catalog manipulations the sync needs to make.

@dedene thanks for sharing that but we should note that the workaround you described is **extremely insecure**. `service:write` grants anyone who has it the ability to "become" that service. Explicitly in connect that means you can get a certificate signed as that service and then access anything that service is allowed to access, but even without Connect it implicitly creates the same risk - you can register an instance of that service which means you can hijack at least part of it's traffic. Granting `service:write` to _any_ service to anonymous is therefore very risky but to all possible services effectively bypasses ACLs entirely for service security!

**Comment by banks** [comment]
I'm leaving this open as we should probably make this clearer in the docs as well as making it generally more obvious how to run this properly with ACLs. That is a known issue before this but this is a worthwhile reminder.

**Comment by andoriyu** [comment]
> @arctiqtim the policy you need for two way sync should be:
> 
> ```
> # read all nodes allows the sync process to _discover_ services to sync them 
> # _to_ k8s since it needs read on the node that they are one to be able to "see" them
> node_prefix "" {
>   policy = "read"
> }
> 
> # write on all services allows the sync process to create service registrations for
> # anything it finds in k8s. If you are using -k8s-service-prefix then you can lock this 
> # down further by only granting write for that specific prefix which is safer since 
> # service:write grants access to the _identity_ of that service and hence anything it has
> # access to when using Connect.
> service_prefix "" {
>   policy = "write"
> }
> ```
> 
> I've not actually tested that so there might be some other subtelty but those are the main catalog manipulations the sync needs to make.
> 
> @dedene thanks for sharing that but we should note that the workaround you described is **extremely insecure**. `service:write` grants anyone who has it the ability to "become" that service. Explicitly in connect that means you can get a certificate signed as that service and then access anything that service is allowed to access, but even without Connect it implicitly creates the same risk - you can register an instance of that service which means you can hijack at least part of it's traffic. Granting `service:write` to _any_ service to anonymous is therefore very risky but to all possible services effectively bypasses ACLs entirely for service security!

Well, yes. and if consul-k8s supported setting token for requests it makes this wouldn't be an issue?

**Comment by dedene** [comment]
@andoriyu Thanks for the quick answer. I was indeed aware of that but we are using it in a very small controlled use case without Connect. But you are absolutely right about the implications for Consul service running with a publicly accessible API. 

I actually ran into this due to the other issue described in #40. While with the 

```
node_prefix "" {
  policy = "read"
}
service_prefix "" {
  policy = "write"
}
```

policy, I observe that services from k8s were getting synced. But I was getting continuously `"Catalog.Deregister" RPC failed to server 10.2.64.19:8300: rpc error making call: rpc error making call: Permission denied` errors. I'm suspecting these were coming from the anti-entropy inside Consul. After experimenting I found I could also not unregister a service manually, even using a master token. Only when adding write access to the anonymous token (`-id 00000000-0000-0000-0000-000000000002`) these errors went away.

**Comment by adilyse** [comment]
@dedene The rpc errors were likely due to anti-entropy. The catalog sync pod registers services directly with the servers, but because of a quirk about the node details provided in the registration, the agents would then unregister the services since they didn't own them. This behavior was fixed in #63 so shouldn't continue to happen.

The sync permissions may not have appeared to fix this policy because it would have been applied to the sync pod, but it was the clients that were attempting the call. If they had not been given any of these same permissions, it was rightly denied. 

Side note to all of this: If you are using the catalog sync provided in the Helm chart without using Connect, it is not strictly necessary to be running the client pods in Kubernetes.

**Comment by adilyse** [comment]
@andoriyu Support for adding an ACL token to the catalog sync process through the Helm chart will be added shortly (consul-helm PR [111](https://github.com/hashicorp/consul-helm/pull/111)) and will be available in the next release. This should unblock ACL users, but we understand that there is still a lot of room for improvement to the user experience when setting ACLs up in Kubernetes. It's something we're aware of and is in process, though it will take a bit longer.

**Comment by adilyse** [comment]
Closing this, as I believe most of the points raised here have been addressed. Please feel free to file a new issue if there are further issues related to this!
